src/sdam/server.ts

```diff
@@ -335,49 +311,64 @@ export class Server extends TypedEventEmitter<ServerEvents> {
-    //       check.
-    if (this.loadBalanced && session && conn == null && isPinnableCommand(cmd, session)) {
-      this.pool.checkOut((err, checkedOut) => {
-        if (err || checkedOut == null) {
-          if (callback) return callback(err);
-          return;
-        }
+    //       require special logic to decrement it again, or would double increment. Instead, we
+    //       increment the count after this check.
 
-        session.pin(checkedOut);
-        this.command(ns, cmd, finalOptions, callback);
-      });
-      return;
-    }
+    const connShouldBePinned =
+      this.loadBalanced && session != null && isPinnableCommand(cmd, session);
 
     this.incrementOperationCount();
+    if (!conn) {
+      try {
+        conn = await this.pool.checkOut();
+      } catch (checkoutError) {
+        this.decrementOperationCount();
+        if (!(checkoutError instanceof PoolClearedError)) this.handleError(checkoutError);
+        throw checkoutError;
+      }
+    }
 
-    this.pool.withConnection(
-      conn,
-      (err, conn, cb) => {
-        if (err || !conn) {
-          this.decrementOperationCount();
-          if (!err) {
-            return cb(new MongoRuntimeError('Failed to create connection without error'));
-          }
-          if (!(err instanceof PoolClearedError)) {
-            this.handleError(err);
-          }
-          return cb(err);
-        }
+    if (connShouldBePinned && !session.isPinned) {
+      session.pin(conn);
+    }
 
-        const handler = makeOperationHandler(this, conn, cmd, finalOptions, (error, response) => {
-          this.decrementOperationCount();
-          cb(error, response);
-        });
-        conn.command(ns, cmd, finalOptions).then(
-          r => handler(undefined, r),
-          e => handler(e)
-        );
-      },
-      callback
-    );
+    // Reauth flow
+    try {
+      return await this.executeCommand(conn, ns, cmd, finalOptions);
+    } catch (operationError) {
+      if (
+        operationError instanceof MongoError &&
+        operationError.code === MONGODB_ERROR_CODES.Reauthenticate
+      ) {
+        conn = await this.pool.reauthenticate(conn);
+        return await this.executeCommand(conn, ns, cmd, finalOptions);
+      } else {
+        throw operationError;
+      }
+    } finally {
+      this.decrementOperationCount();
+      if (!connShouldBePinned) {
+        this.pool.checkIn(conn);
+      }
+    }
+  }
+
+  /**
+   * Execute a command, catching and appropriately rewrapping thrown errors and updating internal
+   * state as appropriate
+   * @internal
+   */
+  private async executeCommand(
+    conn: Connection,
+    ns: MongoDBNamespace,
+    cmd: Document,
+    options?: CommandOptions
+  ): Promise<Document> {
+    try {
+      return await conn.command(ns, cmd, options);
+    } catch (commandError) {
+      throw this.decorateAndHandleError(conn, cmd, options, commandError);
+    }
   }
```

Comment:
this helper doesn't seem to do much conceptually.  What is "executeCommand" and how is that different from the existing "command" method?  All it does is use a try-catch to wrap the error.

maybe we could just 1) fix decorateAndHandleError to always throw errors and then 2) get rid of this helper and then chain in the `command()` method:

```
conn.command(ns, cmd, options).then(e => this.decorateAndHandleError(conn, cmd, options, e))
```